File: ios/AdPlay/Services/AdService.swift
import Foundation
import SwiftUI
import WebKit

struct AdCredit {
    var state: GameState
    var progress: PlayerProgress?
}

protocol AdServing {
    func showBoostAd(type: BoostType) async throws -> AdCredit
}

/// Low-level network that reports fill outcome without calling the game API.
protocol AdNetwork {
    func attempt() async -> AdFillResult
}

final class MockAdService: AdServing {
    private let api: APIClient

    init(api: APIClient) {
        self.api = api
    }

    func showBoostAd(type: BoostType) async throws -> AdCredit {
        try await Task.sleep(nanoseconds: 1_200_000_000)
        let (state, progress) = try await api.mockComplete(boostType: type)
        return AdCredit(state: state, progress: progress)
    }
}

final class AdMobNetwork: AdNetwork {
    func attempt() async -> AdFillResult {
        await AdMobRewardedPresenter.shared.present()
    }
}

final class AdsBitvexNetwork: AdNetwork {
    func attempt() async -> AdFillResult {
        let ok = await AdsBitvexPresenter.shared.present()
        return ok ? .earned : .unavailable
    }
}

/// Single network: attempt → credit on earn.
final class NetworkAdService: AdServing {
    private let api: APIClient
    private let network: AdNetwork

    init(api: APIClient, network: AdNetwork) {
        self.api = api
        self.network = network
    }

    func showBoostAd(type: BoostType) async throws -> AdCredit {
        if DebugAdBypass.available, DebugAdBypass.isEnabled {
            return try await MockAdService(api: api).showBoostAd(type: type)
        }
        switch await network.attempt() {
        case .earned:
            let (state, progress) = try await api.mockComplete(boostType: type)
            return AdCredit(state: state, progress: progress)
        case .declined, .unavailable:
            throw APIError.message("Ad not completed")
        }
    }
}

/// AdMob first, then AdsBitvex on no-fill only (not after a declined watch).
final class WaterfallAdService: AdServing {
    private let api: APIClient
    private let networks: [AdNetwork]

    init(api: APIClient, networks: [AdNetwork]) {
        self.api = api
        self.networks = networks
    }

    func showBoostAd(type: BoostType) async throws -> AdCredit {
        if DebugAdBypass.available, DebugAdBypass.isEnabled {
            return try await MockAdService(api: api).showBoostAd(type: type)
        }
        for network in networks {
            switch await network.attempt() {
            case .earned:
                let (state, progress) = try await api.mockComplete(boostType: type)
                return AdCredit(state: state, progress: progress)
            case .declined:
                throw APIError.message("Ad not completed")
            case .unavailable:
                continue
            }
        }
        throw APIError.message("Ad not completed")
    }
}

enum AdServiceFactory {
    static func make(api: APIClient, provider: String) -> AdServing {
        switch provider.lowercased() {
        case "mock":
            #if DEBUG
            return MockAdService(api: api)
            #else
            // Never ship the mock provider in Release — fall through to waterfall.
            return WaterfallAdService(
                api: api,
                networks: [AdMobNetwork(), AdsBitvexNetwork()],
            )
            #endif
        case "admob":
            return NetworkAdService(api: api, network: AdMobNetwork())
        case "adsbitvex_only":
            return NetworkAdService(api: api, network: AdsBitvexNetwork())
        // Older backends returned "adsbitvex"; treat as waterfall so AdMob is tried first.
        case "waterfall", "adsbitvex":
            return WaterfallAdService(
                api: api,
                networks: [AdMobNetwork(), AdsBitvexNetwork()],
            )
        default:
            return WaterfallAdService(
                api: api,
                networks: [AdMobNetwork(), AdsBitvexNetwork()],
            )
        }
    }
}

@MainActor
final class AdsBitvexPresenter: NSObject, WKScriptMessageHandler, WKNavigationDelegate {
    static let shared = AdsBitvexPresenter()

    private var continuation: CheckedContinuation<Bool, Never>?
    private var host: UIViewController?

    func present() async -> Bool {
        await AppTracking.requestIfNeeded()
        return await withCheckedContinuation { cont in
            self.continuation?.resume(returning: false)
            self.continuation = cont
            presentWebAd()
        }
    }

    private func presentWebAd() {
        guard let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let root = scene.windows.first(where: { $0.isKeyWindow })?.rootViewController
        else {
            finish(false)
            return
        }

        let presenter = root.presentedViewController ?? root
        let web = WKWebView(frame: .zero)
        web.navigationDelegate = self
        web.configuration.userContentController.add(self, name: "AdPlayBridge")
        web.isOpaque = false
        web.backgroundColor = .black
        web.scrollView.backgroundColor = .black

        let vc = UIViewController()
        vc.view.backgroundColor = .black
        web.translatesAutoresizingMaskIntoConstraints = false
        vc.view.addSubview(web)
        NSLayoutConstraint.activate([
            web.topAnchor.constraint(equalTo: vc.view.topAnchor),
            web.bottomAnchor.constraint(equalTo: vc.view.bottomAnchor),
            web.leadingAnchor.constraint(equalTo: vc.view.leadingAnchor),
            web.trailingAnchor.constraint(equalTo: vc.view.trailingAnchor),
        ])
        vc.modalPresentationStyle = .fullScreen
        host = vc
        presenter.present(vc, animated: true)

        let html = Self.htmlPage(appId: "000241")
        web.loadHTMLString(html, baseURL: URL(string: "https://sdk.adsbitvex.com/"))
    }

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage,
    ) {
        guard message.name == "AdPlayBridge" else { return }
        let body = message.body as? [String: Any]
        let status = body?["status"] as? String ?? ""
        finish(status == "ok")
    }

    private func finish(_ ok: Bool) {
        guard let cont = continuation else { return }
        continuation = nil
        if let host {
            host.dismiss(animated: true) {
                self.host = nil
                cont.resume(returning: ok)
            }
        } else {
            cont.resume(returning: ok)
        }
    }

    private static func htmlPage(appId: String) -> String {
        """
        <!DOCTYPE html>
        <html>
        <head>
          <meta charset="utf-8"/>
          <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1"/>
          <script src="https://sdk.adsbitvex.com/functions/v1/ad-script?appid=\(appId)"></script>
        </head>
        <body style="margin:0;background:#000;color:#ccc;font-family:sans-serif;
          display:flex;align-items:center;justify-content:center;height:100vh;text-align:center">
          <p id="status">Loading ad…</p>
          <script>
            function notify(ok, msg) {
              try {
                window.webkit.messageHandlers.AdPlayBridge.postMessage({
                  status: ok ? 'ok' : 'err',
                  message: msg || ''
                });
              } catch (e) {}
            }
            function run(tries) {
              if (typeof window.showadsbitvex === 'function') {
                document.getElementById('status').textContent = 'Watch to unlock boost…';
                window.showadsbitvex()
                  .then(function () { notify(true, ''); })
                  .catch(function (e) {
                    notify(false, (e && e.message) ? e.message : String(e));
                  });
                return;
              }
              if (tries > 50) {
                document.getElementById('status').textContent = 'Ad SDK failed to load';
                notify(false, 'SDK timeout');
                return;
              }
              setTimeout(function () { run(tries + 1); }, 200);
            }
            run(0);
          </script>
        </body>
        </html>
        """
    }
}

File: android/app/src/main/java/com/adplay/app/data/AdService.kt
package com.adplay.app.data

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.adplay.app.BuildConfig
import com.adplay.app.ads.AdFillResult
import com.adplay.app.ads.AdMobRewarded
import com.adplay.app.ads.AdsBitvexAdActivity
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.delay
import kotlinx.coroutines.withTimeout
import kotlinx.coroutines.TimeoutCancellationException

/**
 * Partner abstraction.
 * Waterfall: AdMob rewarded, then AdsBitvex on no-fill.
 * Boost credit via mockCompleteBoost after a successful reward.
 */
interface AdServing {
    suspend fun showBoostAd(type: BoostType): AdCredit
}

private fun interface AdNetwork {
    suspend fun attempt(activity: Activity?): AdFillResult
}

class MockAdService(private val api: ApiClient) : AdServing {
    override suspend fun showBoostAd(type: BoostType): AdCredit {
        delay(1_200)
        return api.mockComplete(type)
    }
}

private class AdMobNetwork : AdNetwork {
    override suspend fun attempt(activity: Activity?): AdFillResult =
        AdMobRewarded.present(activity)
}

private class AdsBitvexNetwork(private val appContext: Context) : AdNetwork {
    override suspend fun attempt(activity: Activity?): AdFillResult {
        val deferred = CompletableDeferred<Boolean>()
        AdsBitvexAdActivity.pendingResult?.cancel()
        AdsBitvexAdActivity.pendingResult = deferred
        val intent = Intent(appContext, AdsBitvexAdActivity::class.java).apply {
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        appContext.startActivity(intent)
        val ok = try {
            withTimeout(180_000) { deferred.await() }
        } catch (_: TimeoutCancellationException) {
            AdsBitvexAdActivity.pendingResult = null
            false
        }
        // WebView partner: treat failure as unavailable so waterfall can continue if more rungs exist.
        return if (ok) AdFillResult.EARNED else AdFillResult.UNAVAILABLE
    }
}

private class NetworkAdService(
    private val api: ApiClient,
    private val appContext: Context,
    private val network: AdNetwork,
) : AdServing {
    override suspend fun showBoostAd(type: BoostType): AdCredit {
        if (DebugAdBypass.isEnabled(appContext)) {
            return MockAdService(api).showBoostAd(type)
        }
        return when (network.attempt(null)) {
            AdFillResult.EARNED -> api.mockComplete(type)
            AdFillResult.DECLINED, AdFillResult.UNAVAILABLE ->
                throw ApiException(408, "Ad not completed")
        }
    }
}

private class WaterfallAdService(
    private val api: ApiClient,
    private val appContext: Context,
    private val networks: List<AdNetwork>,
) : AdServing {
    override suspend fun showBoostAd(type: BoostType): AdCredit {
        if (DebugAdBypass.isEnabled(appContext)) {
            return MockAdService(api).showBoostAd(type)
        }
        for ((index, network) in networks.withIndex()) {
            when (val result = network.attempt(null)) {
                AdFillResult.EARNED -> return api.mockComplete(type)
                AdFillResult.DECLINED -> throw ApiException(408, "Ad not completed")
                AdFillResult.UNAVAILABLE -> {
                    android.util.Log.i("AdPlayAds", "Waterfall rung $index unavailable, trying next")
                    continue
                }
            }
        }
        throw ApiException(408, "Ad not completed")
    }
}

object AdServiceFactory {
    fun make(api: ApiClient, provider: String, context: Context): AdServing {
        val app = context.applicationContext
        val key = provider.lowercase()
        android.util.Log.i("AdPlayAds", "AdServiceFactory provider=$key")
        return when (key) {
            // Never ship the mock provider in release — fall through to waterfall.
            "mock" -> if (BuildConfig.DEBUG) {
                MockAdService(api)
            } else {
                WaterfallAdService(
                    api,
                    app,
                    listOf(AdMobNetwork(), AdsBitvexNetwork(app)),
                )
            }
            "admob" -> NetworkAdService(api, app, AdMobNetwork())
            // Explicit Bitvex-only (skip AdMob). Prefer "waterfall" in production.
            "adsbitvex_only" -> NetworkAdService(api, app, AdsBitvexNetwork(app))
            // "adsbitvex" used to mean Bitvex-only; treat as waterfall so AdMob is tried first
            // even if an older backend still returns adProvider=adsbitvex.
            "waterfall", "adsbitvex" -> WaterfallAdService(
                api,
                app,
                listOf(AdMobNetwork(), AdsBitvexNetwork(app)),
            )
            else -> WaterfallAdService(
                api,
                app,
                listOf(AdMobNetwork(), AdsBitvexNetwork(app)),
            )
        }
    }

    /** Exposed for logs / UI. */
    fun configuredAppId(): String = BuildConfig.ADSBITVEX_APP_ID
}
